R-package/R/xgb.Booster.R

```diff
@@ -226,7 +234,7 @@ xgb.get.handle <- function(object) {
 #' - For normal predictions, the dimension is `[nrows, ngroups]`.
 #' - For `predcontrib=TRUE`, the dimension is `[nrows, ngroups, nfeats+1]`.
 #' - For `predinteraction=TRUE`, the dimension is `[nrows, ngroups, nfeats+1, nfeats+1]`.
-#' - For `predleaf=TRUE`, the dimension is `[nrows, niter, ngroups, num_parallel_tree]`.
+#' - For `predleaf=TRUE`, the dimension is `[nrows, niter * ngroups * num_parallel_tree]`.
```

Comment:
@david-cortes Do you think the `strict_shape` parameter is better? (It makes the prediction output 4-dim array from the C API for predleaf)

Reply:
Yes, I think it'd be nicer to have the output with 3 or more dimensions for leaf predictions from vector-valued trees in `predict.xgboost` - will update it - but I wouldn't say that having arrays with empty dimensions would be preferrable for cases where the output is actually 2D, like binary classification.

For `predict.xgb.Booster`, I'd nevertheless prefer it to stick to what the C interface outputs, so that it'd behave the same as the other language bindings and so that changes from the C side would not require changes in the R logic.

Considering that the next release will be a major version update (3.0), perhaps it could change the logic from the C side to output a 3D array for vector-valued tree leafs instead? It currently does add those extra dimensions for contributions and interactions for example.

Reply:
It's hidden in the `strict_shape` parameter of the `predict` method. It's available for all prediction methods (csr, DMatrix ...). We can make it default for the R package if needed.

But yes, it might output an empty dimension, it's easier to index from my perspective, don't have to worry about the details of the model (I have a fear of array indexing with **conditionally** checking and calling `np.squeeze`, `reshape(-1)`,` np.newaxis` etc, always wish the input for my code to have the deterministic shape).

This is opinionated but I'm open to change. Would love to learn your experience on when varying the dimension based on the model makes things easier or less confusing.

Reply:
I would prefer to keep `strict_shape=FALSE` as the default. I find it confusing to get arrays with dimensions that are not applicable to the given model type that one is using.

`strict_shape=TRUE` would also not be in line with how other packages behave - for example, VGAM determines the output dimensions conditionally on the `family` argument that the model used, returning 1D for binary classification and 2D for multi-class clasification.

Although it looks like reshaping higher-dim arrays to 2D might actually be more common - taking VGAM again as a reference, it squeezes to 2D when predicting `type="terms"` for multinomial logistic, for example, and same for other packages like MGLM. I'd nevertheless still prefer to keep XGBoost's outputs as higher-dim arrays when they represent multiple dimensions.